share/availability/light/availability.go

```diff
@@ -75,62 +77,80 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
 	}
 	defer release()
 
-	// load snapshot of the last sampling errors from disk
 	key := datastoreKeyForRoot(dah)
+	samples := &SamplingResult{}
+
+	// Attempt to load previous sampling results
 	la.dsLk.RLock()
-	last, err := la.ds.Get(ctx, key)
+	data, err := la.ds.Get(ctx, key)
 	la.dsLk.RUnlock()
-
-	// Check for error cases
-	var samples []Sample
-	switch {
-	case err == nil && len(last) == 0:
-		// Availability has already been validated
-		return nil
-	case err != nil && !errors.Is(err, datastore.ErrNotFound):
-		// Other error occurred
-		return err
-	case errors.Is(err, datastore.ErrNotFound):
-		// No sampling result found, select new samples
-		samples = selectRandomSamples(len(dah.RowRoots), int(la.params.SampleAmount))
-	default:
-		// Sampling result found, unmarshal it
-		err = json.Unmarshal(last, &samples)
+	if err != nil {
+		if !errors.Is(err, datastore.ErrNotFound) {
+			return err
+		}
+		// No previous results; create new samples
+		samples = NewSamplingResult(len(dah.RowRoots), int(la.params.SampleAmount))
+	} else {
+		err = json.Unmarshal(data, samples)
 		if err != nil {
 			return err
 		}
+		// Verify total samples count.
+		totalSamples := len(samples.Remaining) + len(samples.Available)
+		if totalSamples != int(la.params.SampleAmount) {
+			return fmt.Errorf("invalid sampling result:"+
+				" expected %d samples, got %d", la.params.SampleAmount, totalSamples)
+		}
+	}
+
+	if len(samples.Remaining) == 0 {
+		// All samples have been processed successfully
+		return nil
 	}
 
 	var (
-		failedSamplesLock sync.Mutex
-		failedSamples     []Sample
+		mutex         sync.Mutex
+		failedSamples []Sample
+		wg            sync.WaitGroup
 	)
 
 	log.Debugw("starting sampling session", "height", header.Height())
-	var wg sync.WaitGroup
-	for _, s := range samples {
+
+	// remove one second from the deadline to ensure we have enough time to process the results
+	samplingCtx, cancel := context.WithCancel(ctx)
+	if deadline, ok := ctx.Deadline(); ok {
+		samplingCtx, cancel = context.WithDeadline(ctx, deadline.Add(-time.Second))
+	}
+	defer cancel()
```

Comment:
❤️

Reply:
Alternatively, we could store results with an independent context to avoid making assumptions about the time it takes. While also running that in a separate routine to respect the real context and return as the user requested.

Reply:
We must wait for store operation to finish before returning to maintain tx properties.